flare/mgp/mgp.py

```diff
@@ -5,117 +5,184 @@
+                        'bounds_3': [[1.2, 1.2, 1.2], [3.5, 3.5, 3.5]],
+                                    # [[lower,lower,lower],[upper,upper,upper]]
+                                    # Values describe lower and upper bounds
+                                    # for the bondlength-bondlength-bondlength
+                                    # grid used to construct and fit 3-body
+                                    # kernels; note that for force MGPs
+                                    # bondlength-bondlength-costheta
+                                    # are the bounds used instead.
+                        'bodies':   [2, 3] # use 2+3 body
+                        'grid_num_2': 64,# Fidelity of the grid
+                        'grid_num_3': [16, 16, 16],# Fidelity of the grid
+                        'svd_rank_2': 64, #Fidelity of uncertainty estimation
+                        'svd_rank_3': 16**3,
+                        'update': True, # if True: accelerating grids
+                                        # generating by saving intermediate
+                                        # coeff when generating grids
+                        'load_grid': None  # Used to load from file
+                        }
     '''
 
-    def __init__(self, hyps, cutoffs, grid_params: dict, struc_params: dict,
-                 mean_only=False, container_only=True, GP=None,
-                 lmp_file_name='lmp.mgp', n_cpus=None, n_sample=100):
-
-        self.hyps = hyps
-        self.cutoffs = cutoffs
-        self.grid_params = grid_params
-        self.struc_params = struc_params
-        self.bodies = grid_params['bodies']
-        self.grid_num_2 = grid_params['grid_num_2']
-        self.bounds_2 = grid_params['bounds_2']
-        self.grid_num_3 = grid_params['grid_num_3']
-        self.bounds_3 = grid_params['bounds_3']
-
-        self.svd_rank_2 = grid_params['svd_rank_2']
-        self.svd_rank_3 = grid_params['svd_rank_3']
-        self.update = grid_params['update']
+    def __init__(self,
+                 grid_params: dict,
+                 struc_params: dict,
+                 map_force: bool=False,
+                 GP: GaussianProcess=None,
+                 mean_only: bool=False,
+                 container_only: bool=True,
+                 lmp_file_name: str='lmp.mgp',
+                 n_cpus: int=None,
+                 n_sample: int=100):
+
+        # load all arguments as attributes
+        self.map_force = map_force
         self.mean_only = mean_only
         self.lmp_file_name = lmp_file_name
         self.n_cpus = n_cpus
         self.n_sample = n_sample
+        self.grid_params = grid_params
+        self.struc_params = struc_params
+        self.hyps_mask = None
+        self.cutoffs = None
+
+        # arg_dict = inspect.getargvalues(inspect.currentframe())[3]
+        # del arg_dict['self'], arg_dict['GP']
+        # self.__dict__.update(arg_dict)
+        self.__dict__.update(grid_params)
```

Comment:
I still have mixed feelings about this convention for loading in grid param values. While it does make the grid params a little uglier to define them explicitly, I do think it's helpful when they are referenced elsewhere, and it also allows IDEs to infer other things about the attributes. I think there's a broader conversation to be had with @YuuuuXie about how perhaps the role of grid_params could be changed-- maybe as a part of the refactory.

Reply:
I agree. As we discuss last time, we will replace all this init with explicit function. That make me wonder, whether we should do the same to hyps_mask, at least for the GP initialization. But that will make everything super lengthy...

Reply:
@stevetorr Absolutely I'm gonna re-organize this part. Actually some parameters in grid_params will be removed. 

@nw13slx If you are gonna change GP, I can wait until you get GP done, and see if we need to follow the same style with MGP